File: ConversationIDE/resources/van_engine_bridge/transcript_parser.py
#!/usr/bin/env python3
import sys, io, json, re
from pathlib import Path
from typing import List, Dict, Any, Optional
from dataclasses import dataclass, field
from datetime import datetime

if sys.platform == 'win32':
    sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')
    sys.stderr = io.TextIOWrapper(sys.stderr.buffer, encoding='utf-8')

@dataclass
class TranscriptMessage:
    role: str
    content: str
    timestamp: Optional[str] = None

@dataclass
class StructuredResponse:
    date: str = ""
    summary: str = ""
    analysis: str = ""
    actions: str = ""
    results: str = ""
    status: str = ""
    next_action: str = ""

@dataclass
class ParsedTranscript:
    raw: str
    last_message: str
    current_response_text: str
    voice_completion: str
    plain_completion: str
    structured: StructuredResponse
    response_state: str
    messages: List[TranscriptMessage] = field(default_factory=list)
    message_count: int = 0
    error: Optional[str] = None

class TranscriptParser:
    def __init__(self, da_name: str = "Assistant", principal_name: str = "User"):
        self.da_name = da_name
        self.principal_name = principal_name

    def parse_file(self, transcript_path: Path) -> ParsedTranscript:
        if not transcript_path.exists():
            return ParsedTranscript(raw="", last_message="", current_response_text="",
                voice_completion="", plain_completion="",
                structured=StructuredResponse(), response_state="error",
                error=f"File not found: {transcript_path}")
        content = transcript_path.read_text(encoding='utf-8')
        return self.parse_content(content)

    def parse_content(self, content: str) -> ParsedTranscript:
        lines = content.strip().split('\n')
        messages = []
        last_assistant_message = ""
        current_response_parts = []
        last_human_index = -1

        for i, line in enumerate(lines):
            if not line.strip():
                continue
            try:
                entry = json.loads(line)
                msg_type = entry.get('type')
                if msg_type == 'user' and self._is_real_user_message(entry.get('message', {})):
                    last_human_index = i
                    messages.append(TranscriptMessage(role='user',
                        content=self._extract_text(entry.get('message', {}).get('content')),
                        timestamp=entry.get('timestamp')))
                elif msg_type == 'assistant':
                    text = self._extract_text(entry.get('message', {}).get('content'))
                    if text:
                        last_assistant_message = text
                        messages.append(TranscriptMessage(role='assistant',
                            content=text, timestamp=entry.get('timestamp')))
            except json.JSONDecodeError:
                continue

        for i in range(last_human_index + 1, len(lines)):
            try:
                entry = json.loads(lines[i])
                if entry.get('type') == 'assistant':
                    text = self._extract_text(entry.get('message', {}).get('content'))
                    if text:
                        current_response_parts.append(text)
            except:
                continue

        current_response_text = '\n'.join(current_response_parts)

        return ParsedTranscript(raw=content, last_message=last_assistant_message,
            current_response_text=current_response_text,
            voice_completion=self._extract_voice_completion(current_response_text),
            plain_completion=self._extract_plain_completion(current_response_text),
            structured=self._extract_structured_sections(current_response_text),
            response_state=self._detect_response_state(last_assistant_message),
            messages=messages, message_count=len(messages))

    def _is_real_user_message(self, message: dict) -> bool:
        content = message.get('content')
        if isinstance(content, str):
            return bool(content.strip())
        if isinstance(content, list):
            for block in content:
                if block.get('type') == 'text' and block.get('text'):
                    return True
        return False

    def _extract_text(self, content: Any) -> str:
        if isinstance(content, str):
            return content
        if isinstance(content, list):
            texts = []
            for block in content:
                if block.get('type') == 'text':
                    texts.append(block.get('text', ''))
                elif 'text' in block:
                    texts.append(str(block['text']))
            return ' '.join(texts)
        return ''

    def _remove_system_reminders(self, text: str) -> str:
        return re.sub(r'<system-reminder>[\s\S]*?</system-reminder>', '', text)

    def _extract_voice_completion(self, text: str) -> str:
        text = self._remove_system_reminders(text)
        patterns = [
            rf'\U0001F5E3\s*\*?{re.escape(self.da_name)}:?\*?\s*(.+?)(?:\n|$)',
            r'\U0001F3AF\s*\*?COMPLETED:?\*?\s*(.+?)(?:\n|$)'
        ]
        for pattern in patterns:
            matches = list(re.finditer(pattern, text, re.IGNORECASE))
            if matches:
                last_match = matches[-1]
                result = last_match.group(1).strip()
                result = re.sub(r'^\[AGENT:\w+\]\s*', '', result)
                return result
        return ""

    def _extract_plain_completion(self, text: str) -> str:
        voice = self._extract_voice_completion(text)
        if voice:
            voice = re.sub(r'\[.*?\]', '', voice)
            voice = re.sub(r'\*\*', '', voice)
            voice = re.sub(r'\*', '', voice)
            voice = re.sub(r'[\U00010000-\U0010ffff]', '', voice)
            voice = re.sub(r'\s+', ' ', voice).strip()
            return voice
        summary_match = re.search(r'\U0001F4CB\s*\*?SUMMARY:?\*?\s*(.+?)(?:\n|$)', text, re.IGNORECASE)
        if summary_match:
            summary = summary_match.group(1).strip()
            return summary[:30] + "..." if len(summary) > 30 else summary
        return ""

    def _extract_structured_sections(self, text: str) -> StructuredResponse:
        text = self._remove_system_reminders(text)
        def extract(pattern: str) -> str:
            match = re.search(pattern, text, re.IGNORECASE)
            return match.group(1).strip() if match else ""
        return StructuredResponse(
            date=extract(r'\U0001F4C5\s*(.+?)(?:\n|$)'),
            summary=extract(r'\U0001F4CB\s*SUMMARY:\s*(.+?)(?:\n|$)'),
            analysis=extract(r'\U0001F50D\s*ANALYSIS:\s*(.+?)(?:\n|$)'),
            actions=extract(r'\u26A1\s*ACTIONS:\s*(.+?)(?:\n|$)'),
            results=extract(r'\u2705\s*RESULTS:\s*(.+?)(?:\n|$)'),
            status=extract(r'\U0001F4CA\s*STATUS:\s*(.+?)(?:\n|$)'),
            next_action=extract(r'\u27A1\uFE0F\s*NEXT:\s*(.+?)(?:\n|$)')
        )

    def _detect_response_state(self, last_message: str) -> str:
        if not last_message:
            return "completed"
        if "AskUserQuestion" in last_message or ("?" in last_message and "COMPLETED" not in last_message):
            return "awaiting_input"
        if re.search(r'error|failed|exception|\u274C|\U0001F6A8', last_message, re.IGNORECASE):
            return "error"
        return "completed"

def main():
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument("transcript_path", help="Path to transcript JSONL file")
    parser.add_argument("--voice", action="store_true", help="Output voice completion")
    parser.add_argument("--plain", action="store_true", help="Output plain completion")
    parser.add_argument("--structured", action="store_true", help="Output structured sections")
    parser.add_argument("--state", action="store_true", help="Output response state")
    args = parser.parse_args()
    parser_obj = TranscriptParser()
    result = parser_obj.parse_file(Path(args.transcript_path))
    if args.voice:
        print(result.voice_completion)
    elif args.plain:
        print(result.plain_completion)
    elif args.structured:
        print(json.dumps(result.structured.__dict__, indent=2))
    elif args.state:
        print(result.response_state)
    else:
        print(json.dumps({
            "last_message": result.last_message[:200],
            "voice_completion": result.voice_completion,
            "response_state": result.response_state,
            "message_count": result.message_count
        }, indent=2))

if __name__ == "__main__":
    main()

File: engine/VanEngine.Core/Services/TranscriptParser.cs
using System.Text.Json;
using System.Text.RegularExpressions;

namespace VanEngine.Core.Services;

public class TranscriptParser
{
    private readonly string _transcriptPath;
    private readonly string _daName;

    public TranscriptParser(string transcriptPath, string daName = "Assistant")
    {
        _transcriptPath = transcriptPath;
        _daName = daName;
    }

    public async Task<ParsedTranscript> ParseAsync()
    {
        if (!File.Exists(_transcriptPath))
            return new ParsedTranscript { Raw = "", Error = "Transcript file not found" };

        var raw = await File.ReadAllTextAsync(_transcriptPath);
        var lines = raw.Split('\n', StringSplitOptions.RemoveEmptyEntries);

        var messages = new List<TranscriptMessage>();
        var lastAssistantMessage = "";
        var currentResponseParts = new List<string>();
        var lastHumanIndex = -1;

        for (int i = 0; i < lines.Length; i++)
        {
            try
            {
                var entry = JsonSerializer.Deserialize<TranscriptEntry>(lines[i]);
                if (entry == null) continue;

                if (entry.Type == "user" && IsRealUserMessage(entry.Message?.Content))
                {
                    lastHumanIndex = i;
                    messages.Add(new TranscriptMessage { Role = "user", Content = ExtractText(entry.Message?.Content), Timestamp = entry.Timestamp });
                }
                else if (entry.Type == "assistant")
                {
                    var text = ExtractText(entry.Message?.Content);
                    if (!string.IsNullOrEmpty(text))
                    {
                        lastAssistantMessage = text;
                        messages.Add(new TranscriptMessage { Role = "assistant", Content = text, Timestamp = entry.Timestamp });
                    }
                }
            }
            catch { }
        }

        for (int i = lastHumanIndex + 1; i < lines.Length; i++)
        {
            try
            {
                var entry = JsonSerializer.Deserialize<TranscriptEntry>(lines[i]);
                if (entry?.Type == "assistant")
                {
                    var text = ExtractText(entry.Message?.Content);
                    if (!string.IsNullOrEmpty(text))
                        currentResponseParts.Add(text);
                }
            }
            catch { }
        }

        var currentResponseText = string.Join("\n", currentResponseParts);
        var voiceCompletion = ExtractVoiceCompletion(currentResponseText, _daName);
        var plainCompletion = ExtractPlainCompletion(currentResponseText, _daName);
        var structured = ExtractStructuredSections(currentResponseText);
        var responseState = DetectResponseState(lastAssistantMessage);

        return new ParsedTranscript
        {
            Raw = raw,
            LastMessage = lastAssistantMessage,
            CurrentResponseText = currentResponseText,
            VoiceCompletion = voiceCompletion,
            PlainCompletion = plainCompletion,
            Structured = structured,
            ResponseState = responseState,
            Messages = messages,
            MessageCount = messages.Count
        };
    }

    private static bool IsRealUserMessage(object? content)
    {
        if (content == null) return false;
        if (content is string str) return !string.IsNullOrWhiteSpace(str);
        if (content is JsonElement elem)
        {
            if (elem.ValueKind == JsonValueKind.String)
                return !string.IsNullOrWhiteSpace(elem.GetString());
            if (elem.ValueKind == JsonValueKind.Array)
            {
                foreach (var item in elem.EnumerateArray())
                {
                    if (item.TryGetProperty("type", out var type) && type.GetString() == "text")
                        return true;
                }
            }
        }
        return false;
    }

    private static string ExtractText(object? content)
    {
        if (content == null) return "";
        if (content is string str) return str;
        if (content is JsonElement elem)
        {
            if (elem.ValueKind == JsonValueKind.String)
                return elem.GetString() ?? "";
            if (elem.ValueKind == JsonValueKind.Array)
            {
                var texts = new List<string>();
                foreach (var item in elem.EnumerateArray())
                {
                    if (item.TryGetProperty("type", out var type))
                    {
                        if (type.GetString() == "text" && item.TryGetProperty("text", out var text))
                            texts.Add(text.GetString() ?? "");
                        else if (item.TryGetProperty("text", out var directText))
                            texts.Add(directText.GetString() ?? "");
                    }
                }
                return string.Join(" ", texts);
            }
        }
        return "";
    }

    private static string ExtractVoiceCompletion(string text, string daName)
    {
        text = RemoveSystemReminders(text);

        var patterns = new[]
        {
            new Regex(@"[\u{1f5e3}\u{fe0f}]\s*\*?" + Regex.Escape(daName) + @":?\*?\s*(.+?)(?:\n|$)", RegexOptions.IgnoreCase),
            new Regex(@"\u{1f3af}\s*\*?COMPLETED:?\*?\s*(.+?)(?:\n|$)", RegexOptions.IgnoreCase)
        };

        foreach (var pattern in patterns)
        {
            var matches = pattern.Matches(text);
            if (matches.Count > 0)
            {
                var lastMatch = matches[^1];
                if (lastMatch.Success)
                {
                    var result = lastMatch.Groups[1].Value.Trim();
                    result = Regex.Replace(result, @"^\[AGENT:\w+\]\s*", "");
                    return result;
                }
            }
        }
        return "";
    }

    private static string ExtractPlainCompletion(string text, string daName)
    {
        var voice = ExtractVoiceCompletion(text, daName);
        if (!string.IsNullOrEmpty(voice))
        {
            voice = Regex.Replace(voice, @"\[.*?\]", "");
            voice = Regex.Replace(voice, @"\*\*?", "");
            voice = Regex.Replace(voice, @"\p{So}", "");
            voice = Regex.Replace(voice, @"\s+", " ").Trim();
            return voice;
        }

        var summaryMatch = Regex.Match(text, @"\u{1f4cb}\s*\*?SUMMARY:?\*?\s*(.+?)(?:\n|$)", RegexOptions.IgnoreCase);
        if (summaryMatch.Success)
        {
            var summary = summaryMatch.Groups[1].Value.Trim();
            return summary.Length > 30 ? summary[..27] + "..." : summary;
        }
        return "";
    }

    private static StructuredResponse ExtractStructuredSections(string text)
    {
        text = RemoveSystemReminders(text);
        return new StructuredResponse
        {
            Date = ExtractSection(text, @"\u{1f4c5}\s*(.+?)(?:\n|$)"),
            Summary = ExtractSection(text, @"\u{1f4cb}\s*SUMMARY:\s*(.+?)(?:\n|$)", true),
            Analysis = ExtractSection(text, @"\u{1f50d}\s*ANALYSIS:\s*(.+?)(?:\n|$)", true),
            Actions = ExtractSection(text, @"\u{26a1}\s*ACTIONS:\s*(.+?)(?:\n|$)", true),
            Results = ExtractSection(text, @"\u{2705}\s*RESULTS:\s*(.+?)(?:\n|$)", true),
            Status = ExtractSection(text, @"\u{1f4ca}\s*STATUS:\s*(.+?)(?:\n|$)", true),
            Next = ExtractSection(text, @"\u{27a1}\u{fe0f}\s*NEXT:\s*(.+?)(?:\n|$)", true)
        };
    }

    private static string ExtractSection(string text, string pattern, bool multiline = false)
    {
        var options = multiline ? RegexOptions.Multiline : RegexOptions.None;
        var match = Regex.Match(text, pattern, options);
        return match.Success ? match.Groups[1].Value.Trim() : "";
    }

    private static string RemoveSystemReminders(string text)
    {
        return Regex.Replace(text, @"<system-reminder>[\s\S]*?<\/system-reminder>", "");
    }

    private static ResponseState DetectResponseState(string lastMessage)
    {
        if (string.IsNullOrEmpty(lastMessage))
            return ResponseState.Completed;

        if (lastMessage.Contains("AskUserQuestion") || (lastMessage.Contains("?") && !lastMessage.Contains("COMPLETED")))
            return ResponseState.AwaitingInput;

        if (Regex.IsMatch(lastMessage, @"error|failed|exception|\u{274c}|\u{1f6a8}", RegexOptions.IgnoreCase))
            return ResponseState.Error;

        return ResponseState.Completed;
    }
}

public enum ResponseState { AwaitingInput, Completed, Error }

public class ParsedTranscript
{
    public string Raw { get; set; } = "";
    public string LastMessage { get; set; } = "";
    public string CurrentResponseText { get; set; } = "";
    public string VoiceCompletion { get; set; } = "";
    public string PlainCompletion { get; set; } = "";
    public StructuredResponse Structured { get; set; } = new();
    public ResponseState ResponseState { get; set; }
    public List<TranscriptMessage> Messages { get; set; } = new();
    public int MessageCount { get; set; }
    public string? Error { get; set; }
}

public class StructuredResponse
{
    public string Date { get; set; } = "";
    public string Summary { get; set; } = "";
    public string Analysis { get; set; } = "";
    public string Actions { get; set; } = "";
    public string Results { get; set; } = "";
    public string Status { get; set; } = "";
    public string Next { get; set; } = "";
}

public class TranscriptMessage
{
    public string Role { get; set; } = "";
    public string Content { get; set; } = "";
    public string? Timestamp { get; set; }
}

public class TranscriptEntry
{
    public string? Type { get; set; }
    public MessageContent? Message { get; set; }
    public string? Timestamp { get; set; }
}

public class MessageContent
{
    public object? Content { get; set; }
    public string? Role { get; set; }
}
